Refactor online lookup flows in song select

bdach opened the pull request:
RFC

## [Refactor realm populating online beatmap set lookup to not require explicit realm refetch](https://github.com/ppy/osu/commit/bd215ca670ef65d99a63fd6da8df634598c1b17f)

- Alternative to / closes https://github.com/ppy/osu/pull/34723
- Closes https://github.com/ppy/osu/issues/34716

Implements idea from https://github.com/ppy/osu/pull/34723#discussion_r2282247507 wherein the "realm populating lookup component" also returns the state of the local beatmap set model after it's done touching it to avoid a realm refresh.

## [Pull up online lookup to song select level to avoid two components doing the same fetch independently](https://github.com/ppy/osu/commit/f4db2632982757f7bab48f814853becb17184d70)

This is not required for anything but it was annoying me for long enough that when I was touching the same areas anyway I figured I might as well try and get rid. Can drop this from the PR if deemed incorrect direction.

## [Fix song select favourite button potentially showing stale data from (un)favourite request callback](https://github.com/ppy/osu/commit/60b16c7c6909a1bce89710dfa7ac4a70cf5fed19)

Noticed in test failures after previous commit.

bdach committed:
Refactor realm populating online beatmap set lookup to not require explicit realm refetch

bdach committed:
Pull up online lookup to song select level to avoid two components doing the same fetch independently

bdach committed:
Fix song select favourite button potentially showing stale data from (un)favourite request callback

bdach committed:
Attempt to fix test failures

That `SongSelect` change is a bit of a hail mary. A bunch of tests are
failing because of "DummyAPIAccess cannot process this request" and I
cannot either replicate or really understand why. The only thing that
makes some semblance of sense to me is that this is unobserved
exceptions bubbling or something because potentially `Exception` is not
accessed if the scheduled delegate never runs.

peppy commented on osu.Game.Tests/Visual/SongSelectV2/TestSceneBeatmapLeaderboardSorting.cs:
```diff
@@ -42,7 +43,8 @@ public partial class TestSceneBeatmapLeaderboardSorting : SongSelectComponentsTe
         private LeaderboardManager leaderboardManager = null!;
-        private RealmPopulatingOnlineLookupSource lookupSource = null!;
+
+        private readonly IBindable<RealmPopulatingOnlineLookupSource.BeatmapSetLookupResult?> onlineLookupResult = new Bindable<RealmPopulatingOnlineLookupSource.BeatmapSetLookupResult?>();
```
Need to do away with subclassing for code legibility 🙈

peppy commented:
I'm almost on board with this, except for one thing: `Local` is only used in [one place](https://github.com/ppy/osu/blob/f4db2632982757f7bab48f814853becb17184d70/osu.Game/Screens/SelectV2/BeatmapMetadataWedge.cs#L397). Could this not potentially be replaced with something like `result.Online?.Metadata.UserTags ?? beatmapSet.Metadata.UserTags`? ie use the online response if available (should always be more correct if non-null) else use local.

This would simplify the result to a single `APIBeatmapSet` if it can work. But I may be missing something.

peppy reviewed (commented):
As commented.

peppy commented:
I do wonder if that case could just be handled with some brief local logic (hash check or otherwise)...

bdach committed:
Fix using wrong status when updating status pill in title wedge

The set's status and the particular beatmap's status are not necessarily
the same. In this context it is the particular beatmap's status that
should be displayed, not the entire set's.

bdach commented:
> some brief local logic (hash check or otherwise)...

It's not clear to me what that means. Which hash would be the one checked here?

And remember, no separate realm refetches allowed, because if you try and do that, we're back at square one of needing `.Refresh()` to guarantee that any latest async changes can be seen.
